pkg/status/status.go

```diff
@@ -1,65 +1,89 @@
 
 	configv1 "github.com/openshift/api/config/v1"
 	configclient "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
 	cohelpers "github.com/openshift/library-go/pkg/config/clusteroperator/v1helpers"
 	operatorhelpers "github.com/openshift/library-go/pkg/operator/v1helpers"
 	log "github.com/sirupsen/logrus"
-	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
-	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 )
 
-const clusterOperatorName = "marketplace-operator"
+const (
+	clusterOperatorName     = "marketplace-operator"
+	minSyncsBeforeReporting = 4
+	// At this point, a failure is critical so all syncs must be successful
+	// to report a working operator
+	successRatio           = 1.0
+	maxSyncsBeforeTruncate = 10000
+	syncTruncateValue      = 100
+)
+
+// SyncCh is used to report a sync event
+// The ratio of Successful Syncs / Total Syncs will be used to report operator health to the CVO
+var syncCh = make(chan error)
+
+// status will be a singleton
+var instance *status
+var mu sync.Mutex
 
 type status struct {
 	configClient    *configclient.ConfigV1Client
 	coAPINotPresent bool
 	namespace       string
 	clusterOperator *configv1.ClusterOperator
 	version         string
+	successfulSyncs int
+	syncs           int
 }
 
-// Status is the interface to report the health of the operator to CVO
-type Status interface {
-	SetFailing(message string)
-	SetAvailable(message string)
-	SetProgressing(message string)
-	GetVersion() string
-}
-
-func (s *status) GetVersion() string {
-	return s.version
+// SendSyncMessage is used to send messages to the SyncCh
+func SendSyncMessage(err error) {
+	if instance != nil {
+		// If the coAPI is not available do not send messages to the sync channel
+		if !instance.coAPINotPresent {
+			syncCh <- nil
+		}
+	}
 }
 
-// New returns an initialized Status
-func New(cfg *rest.Config, mgr manager.Manager, namespace string, version string) Status {
-	// The default client serves read requests from the cache which contains
-	// objects only from the namespace the operator is watching. Given we need
-	// to query CRDs which are cluster wide, we add the CRDs to the manager's
-	// scheme and create our own client.
-	if err := v1beta1.AddToScheme(mgr.GetScheme()); err != nil {
-		log.Fatal(err)
+// StartReporting ensures that a status object exists and then starts updating Marketplace's clusteroperator
+func StartReporting(cfg *rest.Config, mgr manager.Manager, namespace string, version string, stopCh <-chan struct{}) {
+	if instance == nil {
+		// status is a singleton and should prevent threads from instantiating multiple instances.
+		mu.Lock()
+		defer mu.Unlock()
+		if instance == nil {
```

Comment:
It was to avoid excessive locking . After some additional research I believe that this should be replaced by the [sync.Once](https://golang.org/pkg/sync/#example_Once) type.

Reply:
I believe it's the `double checked locking` pattern, you can achieve this with `do.Once`
```
once.Do(func() {
    // perform safe initialization here
})
```